### PR TITLE
perf(gatsby): use fastq instead of better-queue + refactor

### PR DESCRIPTION
**Pull request by vladar**
## Description

This PR is a polished version of this commit: https://github.com/gatsbyjs/gatsby/commit/78d6986069354ed2904ebbcdbd05cd42a3f26ce0

Original comment from @KyleAMathews:

> Various optimizations to query running that make QPS 20% faster. I moved all work into the queue so the event loop wouldn't get overwhelmed for large sites + moved to a more efficient queue https://www.npmjs.com/package/fastq — our previous queue, better-queue was ~7-10% of the CPU by itself and fastq is maybe 0.1%.

One major benefit of `better-queue` was support for assigning priority to tasks. We used it in `develop` to run the currently active query first. But after "query on demand" this is no longer needed as we run only active queries anyway.

## Benchmarks

Benchmarks with [query-filter-sort site](https://github.com/gatsbyjs/gatsby/tree/master/benchmarks/query-filters-sort):

```
cross-env NUM_NODES=10000 yarn bench
```

Before:
```
success run page queries - 1.962s - 1001/1001 510.13/s
success run page queries - 2.088s - 1001/1001 479.45/s
success run page queries - 1.735s - 1001/1001 576.95/s
success run page queries - 2.056s - 1001/1001 486.87/s
success run page queries - 1.780s - 1001/1001 562.41/s
success run page queries - 1.853s - 1001/1001 540.20/s
success run page queries - 1.676s - 1001/1001 597.24/s
success run page queries - 1.740s - 1001/1001 575.15/s
```

After:
```
success run page queries - 1.573s - 1001/1001 636.54/s
success run page queries - 1.566s - 1001/1001 639.17/s
success run page queries - 1.637s - 1001/1001 611.55/s
success run page queries - 1.627s - 1001/1001 615.29/s
success run page queries - 1.665s - 1001/1001 601.12/s
success run page queries - 1.763s - 1001/1001 567.75/s
success run page queries - 1.566s - 1001/1001 639.24/s
success run page queries - 1.575s - 1001/1001 635.54/s
success run page queries - 1.553s - 1001/1001 644.61/s
```

Interestingly version of `fastq` with promises is slower (has basically the same numbers as current version)

### CODE REVIEW

**Comment by KyleAMathews** [comment]
Good deleted/added code ratio 👍 

Looks like a ~12% speedup for 1000 queries (using my handy one-liner — `split(`\n`).map(l => parseFloat(l.split(` `).slice(-1)[0].split(`/`)[0])).reduce((a,b) => a + b) /  9`)

I [got 20%](https://github.com/gatsbyjs/gatsby/commit/78d6986069354ed2904ebbcdbd05cd42a3f26ce0) but that was with 100k queries — the more queries the bigger the speedup I'd assume given better-queue's sorting logic would get slower on bigger datasets.

**Comment by vladar** [comment]
Yeah, also benchmark setup is probably different. I believe my benchmark still does a pass of fast filters (since I am not filtering on the `id` field). Filtering by `id` has less overhead and so the perf benefit is more noticeable.

File: packages/gatsby/src/query/index.js [hunk 1]
```diff
@@ -46,18 +58,83 @@ const groupQueryIds = queryIds => {
...
   })
-  return queryQueue.processBatch(queue, queryJobs, activity)
 }
 
-const createStaticQueryJob = (state, queryId) => {
+function createStaticQueryJob(state, queryId) {
   const component = state.staticQueryComponents.get(queryId)
+  if (!component) {
+    return undefined
+  }
```
Comment (pieh) [inline packages/gatsby/src/query/index.js:137]: Was this check not needed before? It doesn't seem to me like changing to `fastq` would require additional check like that?

File: packages/gatsby/src/query/index.js [hunk 2]
```diff
@@ -68,48 +145,55 @@ const createStaticQueryJob = (state, queryId) => {
...
+function createPageQueryJob(state, queryId) {
+  const page = state.pages.get(queryId)
+
+  // Make sure we filter out pages that don't exist. An example is
+  // /dev-404-page/, whose SitePage node is created via
+  // `internal-data-bridge`, but the actual page object is only
+  // created during `gatsby develop`.
+  if (!page) {
+    return undefined
+  }
```
Comment (pieh) [inline packages/gatsby/src/query/index.js:196]: similar question here? this doesn't seem related to `fastq` swap?

Comment (vladar) [reply packages/gatsby/src/query/index.js:196]: Here is what we had before:

https://github.com/gatsbyjs/gatsby/blob/dad0032db4ea24f6923033906e3b29552f05c7ed/packages/gatsby/src/query/index.js#L91-L103

But now we've moved job creation to worker itself and so need to account for this in the createJob function itself.

File: packages/gatsby/src/query/index.js (hunk 1 above)
Comment (vladar) [reply packages/gatsby/src/query/index.js:137]: See below. This is probably not needed for the static query case. But I've added it for consistency with the page query and also because I am going to convert it to TS at some point and it will complain there anyways.

**Review by pieh — APPROVED** [review]
Awesome results. LGTM
